### PR TITLE
MESH-1718/Correctly Prune STO settlement instructions

### PR DESCRIPTION
**Pull request by CJP10**
### modified logic

- `STO::Invest` - Now correctly prunes instructions.

### CODE REVIEW

File: pallets/sto/src/lib.rs [hunk 1]
```diff
@@ -491,6 +493,10 @@ decl_module! {
                 }
             })?;
 
+            if let Some(id) = instruction_to_prune {
```
Comment (Centril) [inline pallets/sto/src/lib.rs:496]: Would be good to add a comment explaining why we're doing this.

File: pallets/settlement/src/lib.rs [hunk 2]
```diff
@@ -1175,7 +1175,7 @@ impl<T: Config> Module<T> {
         Ok(instructions_processed)
     }
 
-    fn prune_instruction(id: InstructionId) {
+    pub fn prune_instruction(id: InstructionId) {
```
Comment (Neopallium) [inline pallets/settlement/src/lib.rs:1178]: Minor but this don't need to be public anymore.

**Review by Neopallium — APPROVED** [review]
LGTM, just that extra `pub`.

**Review by Neopallium — APPROVED** [review]

File: pallets/settlement/src/lib.rs [hunk 3]
```diff
@@ -1533,49 +1533,35 @@ impl<T: Config> Module<T> {
...
-        max_legs_count: u32,
-    ) -> DispatchResult {
-        with_transaction(|| {
-            Self::base_affirm_with_receipts(
+        match receipt {
+            Some(receipt) => Self::base_affirm_with_receipts(
                 origin,
                 id,
-                receipt_details,
+                vec![receipt],
```
Comment (Centril) [inline pallets/settlement/src/lib.rs:1550]: Maybe using `SmallVec`s would be profitable here. Could be something to investigate (let's add a story for that).

**Review by Centril — APPROVED** [review]

**Comment by Centril** [comment]
Let's add changelog to the PR before merging.

File: pallets/settlement/src/lib.rs [hunk 4]
```diff
@@ -1533,49 +1533,35 @@ impl<T: Config> Module<T> {
...
+            None => {
+                Self::base_affirm_instruction(origin, id, portfolios.into_iter(), max_legs_count)?
+            }
+        };
+        let result = Self::execute_settle_on_affirmation_instruction(
+            id,
+            Self::instruction_affirms_pending(id),
+            Self::instruction_details(id).settlement_type,
+        );
+        Self::prune_instruction(id);
```
Comment (maxsam4) [inline pallets/settlement/src/lib.rs:1563]: Optimization: This prune is only needed in the happy path. If settlement fails, the `with_transaction` block should undo the instruction anyway.

**Pull request merged**